---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup 


```{r}

"%&%" <- function(a,b) paste0(a,b)
library("tidyverse")
library("GenomicRanges")
library("viridis")
library("data.table")
library("RColorBrewer")
library("gridExtra")

serv.dir <- "/home/jason/science/servers/FUSE5/"
proj.dir <- serv.dir %&% "projects/t2d_classification/"
work.dir <- proj.dir %&% "revamp/"
analysis.dir <- work.dir %&% "analysis_files/"
plot.dir <- work.dir %&% "plots/"

cred.df <- fread(work.dir %&%  "genetic_credible_sets/gencred.txt")
cred.df$Locus.ID <- cred.df$CondID

df <- fread(analysis.dir%&%"classified-loci_weighted_with-shared.txt")
df$loc <- map(df$Locus.ID,function(id){
  strsplit(id,split="_")[[1]][1]
}) %>% as.character(.)


```


## Identify heterogeneious loci 


```{r}

get_het_status <- function(hetloc){
  diff.rank <- FALSE
  diff.assign20 <- FALSE
  iter<-1
  sub <- filter(df,loc==hetloc)
  enough.classified <- length(sub$assigned_20[sub$assigned_20!="unclassified"]) >= 2
  for (i in 1:(dim(sub)[1]-1)){
    a <- sub[iter,2:5] %>% as.numeric(.)
    b <- sub[(iter+1),2:5] %>% as.numeric(.)
    evalu <- all(rank(a)==rank(b))
    diff.rank <- ifelse(diff.rank==TRUE,TRUE,
                        ifelse(evalu==FALSE,TRUE,FALSE))
    a <- sub[iter,]$assigned_20
    b <- sub[iter+1,]$assigned_20
    diff.assign20 <- ifelse(diff.assign20==TRUE,TRUE,
                        ifelse(a!=b,TRUE,FALSE))
    iter <- iter + 1 
  }
  return(list(diff.rank,diff.assign20,enough.classified))
}


build_het_df <- function(){
  ids <- unique(cred.df$Locus.ID)
  het.vec <- c()
  for (id in ids){
    l <- strsplit(id,split="_")[[1]]
    if (l[[2]]!="1"){
      het.vec <- append(l[[1]],het.vec)
    }
  }
  het.vec <- het.vec %>% unique(.) # 84 heterogeneous signals 
  out.df <- c()
  for (het in het.vec){
    l <- get_het_status(het)
    build.df <- data.frame(Locus=het,Ranks.differ=l[[1]],
                           Assign.differ=l[[2]],
                           Enough.classified=l[[3]],stringsAsFactors = F)
    out.df <- rbind(out.df,build.df)
  }  
  return(out.df)
}

```


```{r}

het.df <- build_het_df()
het.df$genes <- map(het.df$Locus,function(hetloc){
  filter(df,loc==hetloc)$symbol %>% unique(.) %>% paste0(.,collapse = ",")
}) %>% as.character(.)

sum(het.df$Enough.classified==TRUE)

sub.df <- filter(het.df,Enough.classified==TRUE)

sub.df$Assign.differ %>% sum(.) / length(sub.df$Assign.differ) # 40/57 (70%) 
sub.df$Ranks.differ %>% sum(.) / length(sub.df$Ranks.differ) # 55/57 (96%) with differing rank orders 

```

There are 84 loci with allelic heterogeneity 
57 have enough signals classified at the 0.2 threshold to enable comparison (i.e. two or more)
Of these 57, 40 (70%) have different tissue classifications (heterogeneous assignments) 
Moreover, 55 (96%) have different rank orders (tissue profiles) 


# Plotting function 

```{r}

df$lead.rsid <- map(df$Locus.ID,function(id){
  filter(cred.df,Locus.ID==id)$lead.rsid %>% unique(.)
}) %>% as.character(.)
df$symbol <- map(df$Locus.ID,function(id){
  filter(cred.df,Locus.ID==id)$symbol %>% unique(.)
}) %>% as.character(.)
df$max.ppa <- map(df$Locus.ID,function(id){
  filter(cred.df,Locus.ID==id)$PPA %>% max(.)
}) %>% as.numeric(.)


```


```{r}

het_plot <- function(locus.num){
  input.df <- filter(df,loc==locus.num)
  plot.df <- c()
  tiss.vec <- c("islet","muscle","adipose","liver","unclassified")
  for (i in 1:dim(input.df)[1]){
    row.df <- input.df[i,]
    toa.vec <- row.df %>% 
      select(.,one_of("islet","muscle","adipose","liver","unclassified")) %>% as.numeric(.)
    build.df <- data.frame(signal=row.df$lead.rsid,max.ppa=row.df$max.ppa,
                           toa=toa.vec,tissue=tiss.vec,stringsAsFactors = FALSE)
    plot.df <- rbind(plot.df,build.df)
  }
  rs.vec <- (plot.df %>% filter(.,tissue=="islet") %>% arrange(.,desc(toa)))$signal
  plot.df$signal <- factor(plot.df$signal,levels=rs.vec)
  plot.df$tissue <- factor(plot.df$tissue,
                           levels=c("unclassified","liver","muscle","adipose","islet"))
  
  plot.df$signum <-  map(plot.df$signal,function(rs){
    (filter(df,lead.rsid==rs)$Locus.ID %>% strsplit(.,split="_"))[[1]][2]
  }) %>% as.integer(.)
  sub <- select(plot.df,one_of("signal","signum"))
  sub <- sub[!duplicated(sub),]
  sub$signal <- factor(sub$signal,levels=rs.vec)
  
  rs.font <- ifelse((map(rs.vec,function(rs){
    filter(sub,signal==rs)$signum
  }) %>% as.integer(.) )==1,"bold","plain")
  
  plot.df$maxppa50 <- plot.df$max.ppa>0.5 # ifelse((plot.df$max.ppa>0.5)==TRUE,1.5,1)
  #gene <- input.df$symbol %>% unique(.)
  plot.df$gene <- input.df$symbol %>% unique(.)
  p <- ggplot(data=plot.df,aes(x=signal,y=toa)) + 
    geom_col(aes(fill=tissue,color=maxppa50),size=1) + 
    scale_fill_manual(values=c("grey","brown","red","gold1","olivedrab2")) + 
    scale_y_continuous(breaks=seq(0,1,0.2)) + 
    scale_color_manual(values = c("ghostwhite","black")) + 
    coord_cartesian(ylim=c(0,1),expand = F) + 
    facet_wrap(~gene)+
    #ggtitle(gene) + 
    theme_bw() + 
    theme(axis.title = element_blank(),
          legend.position = "none",
          axis.text.x=element_text(angle=30,vjust=0.9,hjust=0.9,
                                   face=rs.font),
          strip.text = element_text(size=12,face="italic"),
          strip.background = element_rect(fill="ghostwhite")) 
}

```




```{r}

library(gridExtra)

a <- het_plot("222")
b <- het_plot("146")

grid.arrange(a,b,nrow=1)

```

