---
title: "PCA and clustering analysis"
author: "Jason Torres"
date: "11/21/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Setup 

```{r}

"%&%" <- function(a,b) paste0(a,b)
library("data.table")
library("tidyverse")
library("GenomicRanges")
library("viridis")
library("gridExtra")

serv.dir <- "/Users/jtorres/FUSE5/"
#serv.dir <- "/Users/jtorres/Google Drive/Science/" 
work.dir <- serv.dir %&% "projects/t2d_classification/method_B/"
file.dir <- work.dir %&% "analysis_files/"
plot.dir <- work.dir %&% "plots/"

```


Read data files 

```{r}

input.df <- fread(file.dir %&% "loci_profile.txt")
loc.ids <- input.df$Locus.ID

```


Filter input data frame 

```{r}

# Change in size of credible set 
df1 <- dplyr::select(input.df,-one_of("Locus.ID","loc.pos","CHR","refseq")) %>% 
  dplyr::select(.,one_of("null.num"),contains("numDelta")) %>% 
  scale(.,center=TRUE,scale=TRUE)

# Change in maxPPA within credible set 
df2 <- dplyr::select(input.df,-one_of("Locus.ID","loc.pos","CHR","refseq")) %>% 
  dplyr::select(.,one_of("null.num"),contains("maxppDelta")) %>% 
  scale(.,center=TRUE,scale=TRUE)

# Change in size of credible set, ignore total number of SNPs in credible set 
df3 <- dplyr::select(input.df,-one_of("Locus.ID","loc.pos","CHR","refseq")) %>% 
  dplyr::select(.,contains("numDelta")) %>% 
  scale(.,center=FALSE,scale=FALSE)

# Change in maxPPA within credible set, ignore total number of SNPs in credible set 
df4 <- dplyr::select(input.df,-one_of("Locus.ID","loc.pos","CHR","refseq")) %>% 
  dplyr::select(.,contains("maxppDelta")) %>% 
  scale(.,center=TRUE,scale=TRUE)


# Change size of credible set AND change in maxPPA within credible set, ignore total number of SNPs in credible set 
df5 <- dplyr::select(input.df,-one_of("Locus.ID","loc.pos","CHR","refseq")) %>% 
  dplyr::select(.,contains("Delta")) %>% 
  scale(.,center=FALSE,scale=FALSE)

```


# PCA 

Centered/Scaled analysis 

```{r}

get_pve_and_prout <- function(df){
  pr.out <- prcomp(df,scale=TRUE)
  pr.out$sdev # standard deviations of each principal compoment
  pr.var <- pr.out$sdev^2 # Getting the variance explained by each PC 
  pve<-pr.var/sum(pr.var)
  cumulsum<- cumsum(pve)
  pve.df <- data.frame(PC=c(1:length(pve)),pve,cumulsum)  
  return(list(pr.out,pve.df))
}

```


```{r}

prout1 <- get_pve_and_prout(df1)[[1]]
prout2 <- get_pve_and_prout(df2)[[1]]
prout3 <- get_pve_and_prout(df3)[[1]]
prout4 <- get_pve_and_prout(df4)[[1]]
prout5 <- get_pve_and_prout(df5)[[1]]

pve.df1 <- get_pve_and_prout(df1)[[2]]
pve.df2 <- get_pve_and_prout(df2)[[2]]
pve.df3 <- get_pve_and_prout(df3)[[2]]
pve.df4 <- get_pve_and_prout(df4)[[2]]
pve.df5 <- get_pve_and_prout(df5)[[2]]

```



Add more columns with plotting information 

```{r}

input.df$Plot.ID <- input.df$Locus.ID %&% ": " %&% input.df$refseq

```



## Proportion of Variance Explained Plot 


```{r}


scree_cumsum_plot <- function(pve.df){
  pltA <- ggplot(data=pve.df,aes(x=PC)) + 
    geom_line(aes(y=pve)) + 
    geom_point(aes(y=pve),size=3,shape=21,fill=rev(viridis(length(pve.df$PC)))) + 
    scale_y_continuous(limits=c(0,1),breaks=seq(0,1,0.1)) + 
    scale_x_continuous(limits=c(0,length(pve.df$PC)),breaks=seq(0,length(pve.df$PC),1)) + 
    theme_light() + 
    coord_cartesian(ylim=c(0,1),xlim=c(0.5,length(pve.df$PC)+0.5),expand=FALSE) + 
    xlab("Principal Component") + ylab("Proportion of Variance Explained") + 
    ggtitle("Variance explained by principal components") + 
    theme(panel.grid.minor = element_blank(),
          axis.title = element_text(size=8),
          plot.title = element_text(size=10))
  
  pltB <- ggplot(data=pve.df,aes(x=PC)) + 
    geom_line(aes(y=cumulsum)) + 
    geom_point(aes(y=cumulsum),size=3,shape=21,fill=rev(viridis(length(pve.df$PC)))) + 
    scale_y_continuous(limits=c(0,1),breaks=seq(0,1,0.1)) +
    scale_x_continuous(limits=c(0,length(pve.df$PC)),breaks=seq(0,length(pve.df$PC),1)) + 
    theme_light() + 
    coord_cartesian(ylim=c(0,1),xlim=c(0.5,length(pve.df$PC)+0.5),expand=FALSE) + 
    xlab("Principal Component") + ylab("Cumulative proportion of variance explained") + 
    ggtitle("Cumulative variance explained by principal components") + 
    theme(panel.grid.minor = element_blank(),
          axis.title = element_text(size=8),
          plot.title = element_text(size=10))

  grid.arrange(pltA,pltB,nrow=2)  
}


```


```{r}

pltA <- scree_cumsum_plot(pve.df1)
ggsave(plot=pltA,filename=plot.dir%&%"pca_scree_cumsum_centered_numDelta_nullNum.png",
       height=6,width=4)
pltB <- scree_cumsum_plot(pve.df2)
ggsave(plot=pltB,filename=plot.dir%&%"pca_scree_cumsum_centered_maxppDelta_nullNum.png",
       height=6,width=4)
pltC <- scree_cumsum_plot(pve.df3)
ggsave(plot=pltC,filename=plot.dir%&%"pca_scree_cumsum_centered_numDelta.png",
       height=6,width=4)
pltD <- scree_cumsum_plot(pve.df4)
ggsave(plot=pltD,filename=plot.dir%&%"pca_scree_cumsum_centered_maxppDelta.png",
       height=6,width=4)
pltE <- scree_cumsum_plot(pve.df5)
ggsave(plot=pltE,filename=plot.dir%&%"pca_scree_cumsum_centered_Deltas.png",
       height=6,width=4)


```


Based on the scree plots will proceed with the uncentered data frame 

## Loading vector proportion plots 

```{r}

library("RColorBrewer")

loading_plot <- function(pr.output){
  numb <- dim(pr.output$rotation)[1]
  pca.prop.df <- pr.output$rotation %>% as.data.frame(.,row.names=FALSE)
  annot.vec <- gsub("_"," ",row.names(pr.output$rotation))
  pca.prop.df <- map(1:dim(pca.prop.df)[2],function(i){
    vec <- abs(as.numeric(pca.prop.df[,i])) # absolute value step 
    vec <- vec / sum(vec)
  }) %>% unlist(.) %>% matrix(.,nrow=numb,ncol=numb,byrow=FALSE) %>% as.data.frame(.)
  
  names(pca.prop.df) <- "PC" %&% 1:dim(pca.prop.df)[2]
  
  reformat.list <- map(1:dim(pca.prop.df)[1],function(i){
    PC <- names(pca.prop.df) %>% gsub("PC","",.) %>% as.integer(.)
    Score <- as.numeric(pca.prop.df[i,])
    #PC <- rep(i,length(Score))
    Annotation <- rep(annot.vec[i],length(Score))
    sub.df <- data.frame(PC,Score,Annotation,stringsAsFactors = FALSE)
  }) 
  
  out.df <- c()
  for (i in 1:length(reformat.list)){
    sub.df <- reformat.list[[i]]
    out.df <- rbind(out.df,sub.df)
  }
  ggplot(data=out.df,aes(x=PC,y=Score,fill=Annotation)) + 
          geom_bar(stat="identity",color="black",size=0.2) + 
          scale_y_continuous(breaks=seq(0,1,0.1)) + 
          scale_x_continuous(breaks=seq(0,length(unique(out.df$PC)),1)) + 
          scale_fill_manual(values=colorRampPalette(brewer.pal(11,                                  "Spectral"))(length(unique(out.df$PC)) )) + 
          ggtitle("PC Loading Scores for Genomic Annotations") + 
          ylab("Loading Proportion")
}

```


```{r}

pltLA <- loading_plot(prout1)
ggsave(plot=pltLA,filename=plot.dir%&%"pca_loading_centered_numDelta_nullNum.png",
       height=6,width=8)
pltLB <- loading_plot(prout2)
ggsave(plot=pltLB,filename=plot.dir%&%"pca_loading_centered_maxppDelta_nullNum.png",
       height=6,width=8)
pltLC <- loading_plot(prout3)
ggsave(plot=pltLC,filename=plot.dir%&%"pca_loading_centered_numDelta.png",
       height=6,width=8)
pltLD <- loading_plot(prout4)
ggsave(plot=pltLD,filename=plot.dir%&%"pca_loading_centered_maxppDelta.png",
       height=6,width=8)
pltLE <- loading_plot(prout5)
ggsave(plot=pltLE,filename=plot.dir%&%"pca_loading_centered_Deltas.png",
       height=6,width=8)


```


## PCA plots 


Compare PC1 vs PC2 plots for different input (subsets)

```{r}

library("ggrepel")

pca_annotated_plot <- function(data.df,pca.out,pcX,pcY,annot){
  pca.df <- as.data.frame(pca.out$x) 
  pca.df$Locus.ID <- loc.ids #data.df$Locus.ID
  data.df <- data.frame(Locus.ID=loc.ids,refseq=input.df$refseq,data.df,stringsAsFactors=FALSE)
  data.df$Plot.ID <- input.df$Locus.ID %&% ": " %&% input.df$refseq

  pca.df <- inner_join(data.df,pca.df,by="Locus.ID")

  index <- match(annot,names(pca.df))
  vec <- pca.df[,index]
  names(vec) <- pca.df$Plot.ID
  keep.vec <- sort(vec,decreasing=TRUE)[1:10]
  keep.names <- names(keep.vec)
  
  sub.df <- filter(pca.df,Plot.ID %in% keep.names)
  
  ggplot(data=pca.df,aes_string(x=pcX,y=pcY)) + 
    geom_point(shape=21,size=2,color="black",
               aes_string(fill=annot)) +
    scale_fill_viridis(option="plasma") + 
    geom_label_repel(data=sub.df,
                     aes_string(x=pcX,y=pcY,label="Plot.ID"),
                     size=1.5,
                     fontface="bold",
                     fill="ghostwhite",
                     box.padding = unit(0.1,"lines"), 
                     point.padding = unit(0.1,"lines"),
                     alpha=0.7,label.size=0.1) + 
    theme_bw()  
}

pca_annotated_plot(df1,prout1,"PC1","PC2","islet.numDelta")
pca_annotated_plot(df2,prout2,"PC1","PC2","islet.maxppDelta")
pca_annotated_plot(df3,prout3,"PC1","PC2","islet.numDelta")
pca_annotated_plot(df4,prout4,"PC1","PC2","islet.maxppDelta")
pca_annotated_plot(df5,prout5,"PC1","PC2","islet.numDelta")



```



# Hierarchical clustering 




```{r}

#library("ggdendro")

#input.df <- fread(file.dir %&% "full_best-joint_ppa-prop.txt")
#df <- dplyr::select(input.df,-one_of("Locus.ID","islet_shared_repressed","distance_tss_0_5000",
                                    "coding","other")) %>% scale(.,center=TRUE,scale=TRUE) %>% as.data.frame(.,stringsAsFactors=TRUE)
#refseq <- map(1:dim(input.df)[1],function(i){
#  loc <- input.df$Locus.ID[i]
#  gene <- filter(block99.df,Locus.ID==loc)$refseq
#}) %>% as.character(.)
#df$refseq <- refseq
#df$Plot.ID <- input.df$Locus.ID %&% ": " %&% df$refseq

#dist.df <- dplyr::select(df,-one_of("refseq","Plot.ID")) %>% as.data.frame(.)
#row.names(dist.df) <- df$Plot.ID

##dist.df <- dist.df[1:60,]
#dist.df <- dist.df[1:406,]

#data.dist <- dist(dist.df)
#plot(hclust(data.dist),main="Complete Linkage",xlab="",ylab="")


#hc <- hclust(dist(dist.df),method = "complete")
##ggdendrogram(hc,rotate=FALSE,size=2)

#dhc <- as.dendrogram(hc)
#ddata <- dendro_data(dhc,type="rectangle")
#lab.df <- label(ddata)

#Max.Annot <- as.character(map(as.character(lab.df$label),function(id){
#  i <- match(id,row.names(dist.df))
#  vec <- dist.df[i,]
#  vecmax <- max(vec)
#  coli <-match(vecmax,vec)
#  maxannot <- names(dist.df)[coli]
#  return(maxannot)
#}))

#lab.df$Max.Annot <- Max.Annot

#p <- ggplot(segment(ddata)) + 
#  geom_segment(aes(x = x, y = y, xend = xend, yend = yend)) + 
#  coord_flip() + 
#  geom_label(data=lab.df,aes(x=x,y=y-4,label=label,fill=Max.Annot),
#             label.size=0.1,size=2) + 
#  scale_y_reverse(expand = c(0.2, 0)) + 
#  scale_fill_brewer(palette="Spectral") + 
#  theme_bw()
##ggsave(plot=p,filename=plot.dir%&%"cluster_scaled_dummy.png",height=15,width=15,limitsize=FALSE)
#ggsave(plot=p,filename=plot.dir%&%"cluster_scaled.png",height=40,width=10,limitsize=FALSE)
```

