---
title: "14.1_eQTL-enrich-revamp-make-inputs.Rmd"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup

```{r}

"%&%" <- function(a,b) paste0(a,b)

library("tidyverse")
library("data.table")
library("Homo.sapiens")

serv.dir1 <- "/Users/jasont/science/servers/FUSE/" 
serv.dir2 <- "/Users/jasont/science/servers/FUSE5/" 
proj.dir <- serv.dir2 %&% "projects/t2d_classification/"
work.dir <- proj.dir %&% "tactical_analyses/revisions/reviewer_1/"
gtex.dir <- serv.dir2 %&% "datasets/GTEx/v7/eqtl/GTEx_Analysis_v7_eQTL/"
out.dir <- work.dir %&% "enrichment_files/eqtls/"

```


GTEx eQTLs

```{r}

liv.df <- fread("cat " %&% gtex.dir %&% "Liver.v7.signif_variant_gene_pairs.txt.gz" %&% " | zmore")
pb <- txtProgressBar(min=0,max=dim(liv.df)[1],style=3)
liv.eqtls <- map(1:length(liv.df$variant_id),function(i){
  setTxtProgressBar(pb,i)
  vec <- liv.df$variant_id[i] %>% strsplit(.,split="_")
  id <- vec[[1]][1] %&% ":" %&% vec[[1]][2]
}) %>% as.character(.) %>% unique(.)

mus.df <- fread("cat " %&% gtex.dir %&% "Muscle_Skeletal.v7.signif_variant_gene_pairs.txt.gz" %&% " | zmore")
pb <- txtProgressBar(min=0,max=dim(mus.df)[1],style=3)
mus.eqtls <- map(1:length(mus.df$variant_id),function(i){
  setTxtProgressBar(pb,i)
  vec <- mus.df$variant_id[i] %>% strsplit(.,split="_")
  id <- vec[[1]][1] %&% ":" %&% vec[[1]][2]
}) %>% as.character(.) %>% unique(.)

adi.sub.df <- fread("cat " %&% gtex.dir %&% "Adipose_Subcutaneous.v7.signif_variant_gene_pairs.txt.gz" %&% " | zmore")
pb <- txtProgressBar(min=0,max=dim(adi.sub.df)[1],style=3)
adi.sub.eqtls <- map(1:length(adi.sub.df$variant_id),function(i){
  setTxtProgressBar(pb,i)
  vec <- adi.sub.df$variant_id[i] %>% strsplit(.,split="_")
  id <- vec[[1]][1] %&% ":" %&% vec[[1]][2]
}) %>% as.character(.) %>% unique(.)

adi.visc.df <- fread("cat " %&% gtex.dir %&% "Adipose_Visceral_Omentum.v7.signif_variant_gene_pairs.txt.gz" %&% " | zmore")
pb <- txtProgressBar(min=0,max=dim(adi.visc.df)[1],style=3)
adi.visc.eqtls <- map(1:length(adi.visc.df$variant_id),function(i){
  setTxtProgressBar(pb,i)
  vec <- adi.visc.df$variant_id[i] %>% strsplit(.,split="_")
  id <- vec[[1]][1] %&% ":" %&% vec[[1]][2]
}) %>% as.character(.) %>% unique(.)

```



Islet eQTLs


```{r}

inspire.dir <- serv.dir2 %&% "datasets/INSPIRE_2019/"

insp.exon.df <- fread(inspire.dir %&% "media-1.txt")
pb <- txtProgressBar(min=0,max=dim(insp.exon.df)[1],style=3)
insp.exon.eqtls <- map(1:dim(insp.exon.df)[1],function(i){
  setTxtProgressBar(pb,i)
  vec <- c(insp.exon.df$SNPchr[i],insp.exon.df$StartSNP[i])
  id <- vec[1] %&% ":" %&% vec[2]
}) %>% as.character(.) %>% unique(.)


insp.gene.df <- fread(inspire.dir %&% "media-2.txt")
pb <- txtProgressBar(min=0,max=dim(insp.gene.df)[1],style=3)
insp.gene.eqtls <- map(1:dim(insp.gene.df)[1],function(i){
  setTxtProgressBar(pb,i)
  vec <- c(insp.gene.df$chrSNP[i],insp.gene.df$StartSNP[i])
  id <- vec[1] %&% ":" %&% vec[2]
}) %>% as.character(.) %>% unique(.)

mvdb.df <- fread(serv.dir2 %&% "datasets/mvdb_eSNP_coords.txt",header=F)
mvdb.eqtls <- mvdb.df$V1 %&% ":" %&% mvdb.df$V2


fastqtl.df <- fread("cat " %&% out.dir %&% "snp_keyfile_fdr05.txt.gz" %&% " | zmore")
pb <- txtProgressBar(min=0,max=dim(fastqtl.df)[1],style=3)
fastqtl.eqtls <- map(1:dim(fastqtl.df)[1],function(i){
  setTxtProgressBar(pb,i)
  vec <- c(fastqtl.df$CHR[i],fastqtl.df$POS[i])
  id <- strsplit(vec[1],split="chr")[[1]][2] %&% ":" %&% vec[2]
}) %>% as.character(.)


```


```{r}

islet.eqtls  <-  fastqtl.eqtls

```

Save tissue-specific eQTL sets 

```{r}

adi.union.eqtls <- c(adi.sub.eqtls,adi.visc.eqtls) %>% unique(.)
isl.spec <- islet.eqtls[!(islet.eqtls %in% c(mus.eqtls,liv.eqtls,adi.union.eqtls))]
mus.spec <- mus.eqtls[!(mus.eqtls %in% c(islet.eqtls,liv.eqtls,adi.union.eqtls))]
liv.spec <- liv.eqtls[!(liv.eqtls %in% c(islet.eqtls,adi.union.eqtls,mus.eqtls))]
adi.union.spec <- adi.union.eqtls[!(adi.union.eqtls %in% c(islet.eqtls,liv.eqtls,mus.eqtls))]
adi.visc.spec <- adi.visc.eqtls[!(adi.visc.eqtls %in% c(islet.eqtls,liv.eqtls,mus.eqtls,adi.sub.eqtls))]
adi.sub.spec <- adi.sub.eqtls[!(adi.sub.eqtls %in% c(islet.eqtls,liv.eqtls,mus.eqtls,adi.visc.eqtls))]

```


```{r}

write.table(x=isl.spec,file=out.dir %&% "islet-specific-esnps.txt",sep="\t",quote=F,row.names=F,col.names=F)
write.table(x=mus.spec,file=out.dir %&% "muscle-specific-esnps.txt",sep="\t",quote=F,row.names=F,col.names=F)
write.table(x=adi.union.spec,file=out.dir %&% "adipose-union-specific-esnps.txt",sep="\t",quote=F,row.names=F,col.names=F)
write.table(x=adi.visc.spec,file=out.dir %&% "adipose-visc-specific-esnps.txt",sep="\t",quote=F,row.names=F,col.names=F)
write.table(x=adi.sub.spec,file=out.dir %&% "adipose-sub-specific-esnps.txt",sep="\t",quote=F,row.names=F,col.names=F)
write.table(x=liv.spec,file=out.dir %&% "liver-specific-esnps.txt",sep="\t",quote=F,row.names=F,col.names=F)



```



