---
title: "Expression specificity scores"
author: "Jason Torres"
date: "June 2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup 

```{r}

"%&%" <- function(a,b) paste0(a,b)

library("tidyverse")
library("data.table")
library("Homo.sapiens")
library("GenomicRanges")
library('rtracklayer')
library("GenABEL")

fuse.path <- "/home/jason/science/servers/"
serv.dir <- fuse.path %&% "FUSE5/"
work.dir <- serv.dir %&% "projects/t2d_classification/"
afile.dir <- work.dir %&% "analysis_files/"
gtex.dir <- serv.dir %&% "datasets/GTEx/v7/"

```

Commented out following section as needed to evaluate different data sets and generate gene length dataframe file only once

```{r}

# Approach for obtaining gene lengths 
gtf.gr <- import(serv.dir %&% "/datasets/gencode.v19.genes.v7.patched_contigs.gtf")
gtf.df <- as.data.frame(gtf.gr,stringsAsFactors=FALSE)

ens.vec <- gtf.df$gene_id %>% unique(.)
pb <- txtProgressBar(min=0,max=length(ens.vec),style=3)
gene.length.v2 <- purrr::map(1:length(ens.vec),function(i){
  setTxtProgressBar(pb,i)
  ens <- ens.vec[i]
  sub <- filter(gtf.df,gene_id==ens,type=="gene")
  #sub.gr <- GRanges(seqnames=sub$chrom,IRanges(sub$cdsStart,sub$cdsEnd),sub$strand)
  #sum(width(reduce(sub.gr)))
  sub.gr <- GRanges(seqnames="chr"%&%sub$seqnames,IRanges(sub$start,sub$end),sub$strand)
  sum(width(reduce(sub.gr)))
}) %>% as.integer(.) 

ensid.v2.df <- data.frame(ensid=ens.vec,gene.length.v2,stringsAsFactors = FALSE)

```

Note of the gene that interect, lengths seem to be mostly concordant, however, the "ENSGR" genes (i.e. lincRNAs) are not present in the v1 ENSEMBLE gene file, so opt for ensid.v2.df for TPM normalization 

```{r}

#write.table(ensid.v2.df,file=afile.dir%&%"gencode.v19.gene_lengths.txt",
#            sep="\t",quote=FALSE,row.names=FALSE)
#ensid.v2.df <- read.table(afile.dir%&%"gencode.v19.gene_lengths.txt",header=TRUE,
#                          stringsAsFactors=FALSE)
```



# Oxford Islets (Unstranded)

```{r}

#vcf.dir <- "/well/mccarthy/projects/118_islets_vandeBunt_plosgen_2015/genotypes/"
# Id file info obtained from header line in any file from vcf.dir 
id.df <- fread(work.dir %&% "input_data/oxford-islet-ids.txt")
id.vec <- id.df$ID
#exp.dir <- "/well/got2d/rna-seq/data/Oxford_islets/unstranded_libraries/"
# Count data copied from exp.dir data above 
unstranded.df <- fread(work.dir %&% "input_data/28.03.2017.unstranded_libraries.gene.counts.tsv")
stranded.df <- fread(work.dir %&% "input_data/17.03.2017.stranded_libraries.gene.counts.tsv")
islet.df <- inner_join(unstranded.df,stranded.df,by="GeneID") %>% 
  dplyr::select(.,one_of("GeneID","GeneName.x",id.vec))

#Warning message:
#Unknown variables: `H262`, `ISL-79`, `R032`, `R036` # Legacy samples, much lower read depths, wouldn't want to use these anyways, use the 114 
# These IDs are not available in the count matrices 

ensids <- islet.df$GeneID
symb.vec <-islet.df$GeneName.x
ensid.vec <- ensids #map(ensids,function(e){strsplit(e,".",fixed=TRUE)[[1]][1]}) %>% 
#as.character(.)
pb <- txtProgressBar(min=0,max=length(ensid.vec),style=3)
gl.vec <- purrr::map(1:length(ensid.vec),function(i){
  setTxtProgressBar(pb,i)
  ens <- ensid.vec[i]
  filter(ensid.v2.df,ensid==ens)$gene.length
}) %>% as.integer(.) 

gl.df <- data.frame(ensembl=ensids,symbol=symb.vec,ensid=ensid.vec,gene.length=gl.vec,
                    stringsAsFactors = FALSE)

```


## TPM Normalization 


```{r}

gl.kb.vec <- gl.vec /1000 # gene lengths in kilo bases 
step1.mat <- dplyr::select(islet.df,-one_of("GeneID","GeneName.x")) %>% as.matrix(.)
row.names(step1.mat) <- islet.df$GeneID
#step2.mat <- (step1.mat / gl.vec) %>% na.omit(.)
step2.mat <- (step1.mat / gl.kb.vec) %>% na.omit(.)

norm.factors <- colSums(step2.mat) / 1e6

for (i in 1:dim(step2.mat)[2]){
  nf <- norm.factors[i]
  step2.mat[,i] <- step2.mat[,i] / nf 
}

pb <- txtProgressBar(min=0,max=length(row.names(step2.mat)),style=3)
GeneName=purrr::map(1:length(row.names(step2.mat)),function(i){
  setTxtProgressBar(pb,i)
  id <- row.names(step2.mat)[i];
  filter(islet.df,GeneID==id)$GeneName.x}) %>% as.character(.)

islet.tpm.df <- cbind(GeneID=row.names(step2.mat),
                      GeneName,
                        as.data.frame(step2.mat))
islet.tpm.df$GeneID <- as.character(islet.tpm.df$GeneID)
islet.tpm.df$GeneName <- as.character(islet.tpm.df$GeneName)

write.table(islet.tpm.df,file=afile.dir%&%"islet_expression_tpm_v1.txt",
            sep="\t",quote=F,row.names=F)
islet.tpm.df <- fread(afile.dir%&%"islet_expression_tpm_v1.txt")

```

# Islet samples (n=174)


```{r}

df174.df <- fread(work.dir %&% "tactical_analyses/input_files/islet_tpm_174.tsv")



```



# GTEx 

Get tissue IDs of interest, read expression count file, create tissue count matrices  

```{r}

samp.df <- fread(gtex.dir %&% "GTEx_v7_Annotations_SampleAttributesDS.txt")
unique(samp.df$SMTSD)
mus.ids <- filter(samp.df,SMTSD=="Muscle - Skeletal")$SAMPID
adi.ids <- filter(samp.df,SMTSD=="Adipose - Subcutaneous")$SAMPID
liv.ids <- filter(samp.df,SMTSD=="Liver")$SAMPID

expr.df <- fread(gtex.dir %&% "expression/GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_reads.gct")
# Note: ran the above lines in the server due to the memory requirements, after subsetting, ran remaining analysis here 

mus.df <- dplyr::select(expr.df,one_of("Name","Description",mus.ids))
adi.df <- dplyr::select(expr.df,one_of("Name","Description",adi.ids))
liv.df <- dplyr::select(expr.df,one_of("Name","Description",liv.ids))
write.table(mus.df,file=gtex.dir%&%"expression/muscle_counts.txt",
            sep="\t",quote=F,row.names=F)
write.table(adi.df,file=gtex.dir%&%"expression/adipose_counts.txt",
            sep="\t",quote=F,row.names=F)
write.table(liv.df,file=gtex.dir%&%"expression/liver_counts.txt",
            sep="\t",quote=F,row.names=F)
mus.df <- fread(gtex.dir%&%"expression/muscle_counts.txt")
adi.df <- fread(gtex.dir%&%"expression/adipose_counts.txt")
liv.df <- fread(gtex.dir%&%"expression/liver_counts.txt")

```

## GTEx TPM Normalization 

```{r}

all(mus.df$Name == adi.df$Name) & all(mus.df$Name ==liv.df$Name) # Check 

ensid.vec <- mus.df$Name
symb.vec <-mus.df$Description

pb <- txtProgressBar(min=0,max=length(ensid.vec),style=3)
gl.vec <- map(1:length(ensid.vec),function(i){
  setTxtProgressBar(pb,i)
  ens <- ensid.vec[i]
  #filter(ensid.df,ensid==ens)$gene.length
  filter(ensid.v2.df,ensid==ens)$gene.length
}) %>% as.integer(.)


```

```{r}

gtex_tpm_normalize <- function(df,gl.vec,tissue.name){
  print(tissue.name);print("")
  gl.kb.vec <- gl.vec /1000 # gene lengths in kilo bases 
  step1.mat <- dplyr::select(df,-one_of("Name","Description")) %>% as.matrix(.)
  row.names(step1.mat) <- df$Name
  step2.mat <- (step1.mat / gl.kb.vec) %>% na.omit(.)
  
  norm.factors <- colSums(step2.mat) / 1e6
  
  for (i in 1:dim(step2.mat)[2]){
    nf <- norm.factors[i]
    step2.mat[,i] <- step2.mat[,i] / nf 
  }
  
  pb <- txtProgressBar(min=0,max=length(row.names(step2.mat)),style=3)
  GeneName=map(1:length(row.names(step2.mat)),function(i){
    setTxtProgressBar(pb,i)
    id <- row.names(step2.mat)[i];
    filter(df,Name==id)$Description}) %>% as.character(.)
  
  tpm.df <- cbind(GeneID=row.names(step2.mat),
                        GeneName,
                          as.data.frame(step2.mat))
  tpm.df$GeneID <- as.character(tpm.df$GeneID)
  tpm.df$GeneName <- as.character(tpm.df$GeneName)
  
  write.table(tpm.df,file=afile.dir%&%tissue.name%&%"_expression_tpm.txt",
              sep="\t",quote=F,row.names=F)
  return(tpm.df)
}

mus.tpm.df <- gtex_tpm_normalize(mus.df,gl.vec,"muscle")
adi.tpm.df <- gtex_tpm_normalize(adi.df,gl.vec,"adipose")
liv.tpm.df <- gtex_tpm_normalize(liv.df,gl.vec,"liver")




```





# RNTransfrom 

```{r}

islet.tpm.df <- fread(afile.dir%&%"islet_expression_tpm.txt")
mus.tpm.df <- fread(afile.dir%&%"muscle"%&%"_expression_tpm.txt")  
adi.tpm.df <- fread(afile.dir%&%"adipose"%&%"_expression_tpm.txt")  
liv.tpm.df <- fread(afile.dir%&%"liver"%&%"_expression_tpm.txt")  

test.df <- summarise(islet.tpm.df,islet=mean("H186":"R051"))

average_tpm_rn <- function(tpm.df){
  out.df <- c()
  pb <- txtProgressBar(min=0,max=dim(tpm.df)[1],style=3)
  for (i in 1:(dim(tpm.df)[1])){
    setTxtProgressBar(pb,i)
    GeneID <- tpm.df$GeneID[i]
    GeneName <- tpm.df$GeneName[i]
    Avg <- tpm.df[i,3:dim(tpm.df)[2]] %>% as.numeric(.) %>% mean(.)
    build.df <- data.frame(GeneID,GeneName,Avg,stringsAsFactors = FALSE)
    out.df <- rbind(out.df,build.df)
  }
  RN <- rntransform(out.df$Avg)
  out.df$RN <- RN
  return(out.df)
}

average_tpm_rr <- function(tpm.df){ # Reverse Rank approach # alternative
  out.df <- c()
  pb <- txtProgressBar(min=0,max=dim(tpm.df)[1],style=3)
  for (i in 1:(dim(tpm.df)[1])){
    setTxtProgressBar(pb,i)
    GeneID <- tpm.df$GeneID[i]
    GeneName <- tpm.df$GeneName[i]
    Avg <- tpm.df[i,3:dim(tpm.df)[2]] %>% as.numeric(.) %>% mean(.)
    build.df <- data.frame(GeneID,GeneName,Avg,stringsAsFactors = FALSE)
    out.df <- rbind(out.df,build.df)
  }
  RN <- rank(out.df$Avg)
  out.df$RN <- RN
  return(out.df)
}

build_rn_df <- function(){
  df1 <- average_tpm_rn(islet.tpm.df)
  df2 <- average_tpm_rn(mus.tpm.df)
  df3 <- average_tpm_rn(adi.tpm.df)
  df4 <- average_tpm_rn(liv.tpm.df)
  df <- inner_join(df1,df2,by="GeneID")
  df <- inner_join(df,df3,by="GeneID")
  df <- inner_join(df,df4,by="GeneID")
  df <- dplyr::select(df,one_of("GeneID","GeneName.x","RN.x","RN.y","RN.x.x","RN.y.y"))
  names(df) <- c("GeneID","GeneName","islet","muscle","adipose","liver")
  return(df)
}

build_rr_df <- function(){
  df1 <- average_tpm_rr(islet.tpm.df)
  df2 <- average_tpm_rr(mus.tpm.df)
  df3 <- average_tpm_rr(adi.tpm.df)
  df4 <- average_tpm_rr(liv.tpm.df)
  df <- inner_join(df1,df2,by="GeneID")
  df <- inner_join(df,df3,by="GeneID")
  df <- inner_join(df,df4,by="GeneID")
  df <- dplyr::select(df,one_of("GeneID","GeneName.x","RN.x","RN.y","RN.x.x","RN.y.y"))
  names(df) <- c("GeneID","GeneName","islet","muscle","adipose","liver")
  return(df)
}

ess_rn_scores <- function(rn.df){
  rn.df$islet <- rn.df$islet 
  rn.df$muscle <- rn.df$muscle  
  rn.df$adipose <- rn.df$adipose 
  rn.df$liver <- rn.df$liver 
  out.df <- c()
  pb <- txtProgressBar(min=0,max=dim(rn.df)[1],style=3)
  for (i in 1:dim(rn.df)[1]){
    setTxtProgressBar(pb,i)
    tot <- dplyr::select(rn.df[i,],-one_of("GeneID","GeneName")) %>% as.numeric(.) %>% sum(.)
    ens <- rn.df$GeneID[i]; ensname <- rn.df$GeneName[i]
    
    if (tot<0){
      islet.score <- 0.25
      muscle.score <- 0.25
      adipose.score <- 0.25
      liver.score <- 0.25     
    } else{
      vec <- dplyr::select(rn.df[i,],-one_of("GeneID","GeneName")) %>%
        as.numeric(.)
      vec[vec<0] <- 0
      tot <- vec %>% sum(.)
      val <- ifelse(rn.df$islet[i]>0,rn.df$islet[i],0)
      islet.score <- val / tot
      val <- ifelse(rn.df$muscle[i]>0,rn.df$muscle[i],0)
      muscle.score <- val / tot
      val <- ifelse(rn.df$adipose[i]>0,rn.df$adipose[i],0)
      adipose.score <- val / tot
      val <- ifelse(rn.df$liver[i]>0,rn.df$liver[i],0)
      liver.score <- val / tot      
    }

    build.df <- data.frame(GeneID=ens,GeneName=ensname,islet.score,muscle.score,
                           adipose.score,liver.score,stringsAsFactors=FALSE)
    out.df <- rbind(out.df,build.df)
  }
  return(out.df)
}

ess_rr_scores <- function(rr.df){
  rr.df$islet <- rr.df$islet 
  rr.df$muscle <- rr.df$muscle 
  rr.df$adipose <- rr.df$adipose 
  rr.df$liver <- rr.df$liver 
  out.df <- c()
  pb <- txtProgressBar(min=0,max=dim(rr.df)[1],style=3)
  for (i in 1:dim(rn.df)[1]){
    setTxtProgressBar(pb,i)
    tot <- dplyr::select(rr.df[i,],-one_of("GeneID","GeneName")) %>% as.numeric(.) %>% sum(.)
    ens <- rr.df$GeneID[i]; ensname <- rr.df$GeneName[i]
    islet.score <- rr.df$islet[i] / tot
    muscle.score <- rr.df$muscle[i] / tot
    adipose.score <- rr.df$adipose[i] / tot
    liver.score <- rr.df$liver[i] / tot
    build.df <- data.frame(GeneID=ens,GeneName=ensname,islet.score,muscle.score,
                           adipose.score,liver.score,stringsAsFactors=FALSE)
    out.df <- rbind(out.df,build.df)
  }
  return(out.df)
}


```


```{r}

rn.df <- build_rn_df()
write.table(rn.df,file=afile.dir%&%
            "rntransform.txt",
            sep="\t",quote=F,row.names=F)
#rn.df <- fread(afile.dir%&%"rntransform.txt")

rr.df <- build_rr_df()
write.table(rr.df,file=afile.dir%&%
            "reverse-rank.txt",
            sep="\t",quote=F,row.names=F)
#rr.df <- fread(afile.dir%&%"reverse-rank.txt")


ess.rn.df <- ess_rn_scores(rn.df)
write.table(ess.rn.df,file=afile.dir%&%"expression_specificity_scores-rntransform.txt",
            sep="\t",quote=F,row.names=F)

```






# Expression specificity scores 


```{r}

all(islet.tpm.df$GeneID %in% mus.tpm.df$GeneID)
all(mus.tpm.df$GeneID %in% islet.tpm.df$GeneID)
all(adi.tpm.df$GeneID == mus.tpm.df$GeneID)
all(liv.tpm.df$GeneID == mus.tpm.df$GeneID)


build_ess_df <- function(){
  ess.df <- c() 
  ens.vec <- islet.tpm.df$GeneID
  pb <- txtProgressBar(min=0,max=length(ens.vec),style=3)
  for (i in 1:length(ens.vec)){
    setTxtProgressBar(pb,i)
    ens <- ens.vec[i]
    ensname <- filter(islet.tpm.df,GeneID==ens)$GeneName
    islet.vec <- filter(islet.tpm.df,GeneID==ens) %>% dplyr::select(.,-one_of("GeneID","GeneName")) %>% as.numeric(.)
    mus.vec <- filter(mus.tpm.df,GeneID==ens) %>% dplyr::select(.,-one_of("GeneID","GeneName")) %>% as.numeric(.)
    adi.vec <- filter(adi.tpm.df,GeneID==ens) %>% dplyr::select(.,-one_of("GeneID","GeneName")) %>% as.numeric(.)
    liv.vec <- filter(liv.tpm.df,GeneID==ens) %>% dplyr::select(.,-one_of("GeneID","GeneName")) %>% as.numeric(.)
    tot <- median(islet.vec) + median(mus.vec) + median(adi.vec) + median(liv.vec) 
    islet.score <- median(islet.vec) / tot 
    muscle.score <- median(mus.vec) / tot 
    adipose.score <- median(adi.vec) / tot 
    liver.score <- median(liv.vec) / tot 
    build.df <- data.frame(GeneID=ens,GeneName=ensname,islet.score,muscle.score,adipose.score,liver.score,stringsAsFactors=FALSE)
    ess.df <- rbind(ess.df,build.df)
  }
  return(ess.df)
}

ess.df <- build_ess_df()

write.table(ess.df,file=afile.dir%&%"expression_specificity_scores.txt",
            sep="\t",quote=F,row.names=F)



```


# Sanity check 

```{r}

ref.isl.df1 <- fread(afile.dir %&% "17.03.2017.stranded_libraries.gene.tpm.tsv")
ref.isl.df2 <- fread(afile.dir %&% "28.03.2017.unstranded_libraries.gene.tpm.tsv")
str(ref.isl.df1)
ref.isl.df <- inner_join(ref.isl.df1,ref.isl.df2,by="GeneID") %>% dplyr::select(.,one_of("GeneID","GeneName.x",id.vec))

g <- "ENSG00000237851.1"
x <- filter(ref.isl.df,GeneID==g) %>% dplyr::select(.,-one_of("GeneID","GeneName","GeneName.x")) %>% as.numeric(.)
y <- filter(islet.tpm.df,GeneID==g) %>% dplyr::select(.,-one_of("GeneID","GeneName","GeneName.x")) %>% as.numeric(.)

g.vec <- islet.tpm.df$GeneID[islet.tpm.df$GeneID %in% ref.isl.df$GeneID] 
pb <- txtProgressBar(min=0,max=1000,style=3)
r2.vec <- map(1:1000,function(i){
  setTxtProgressBar(pb,i)
  g <- g.vec[i]
  x <- filter(ref.isl.df,GeneID==g) %>% dplyr::select(.,-one_of("GeneID","GeneName","GeneName.x")) %>% as.numeric(.)
  y <- filter(islet.tpm.df,GeneID==g) %>% dplyr::select(.,-one_of("GeneID","GeneName","GeneName.x")) %>% as.numeric(.)
  cor.test(x,y)$estimate %>% as.numeric(.)
}) %>% as.numeric()


```



